### PR TITLE
Add support for areas to Razor Pages

### LINKED ISSUE #6926: Use Razor Pages in MVC area
I'm trying to mix Razor Pages with MVC and by that I mean running a Razor Pages "app" inside an MVC area. So if I run the app below I would like `http://localhost:500` to return the standard MVC `HomeController` `Index` action. And if I add the area like this `http://localhost:5000/app` I would like the Razor Pages app to respond. Is this possible?

- Areas
  - App
      - _Layout.cshtml
      - _ViewImports.cshtml
      - _ViewStart.cshtml
      - Index.cshtml
      - Index.cshtml.cs
- Controllers
- Models
- Views
- wwwroot
- Program.cs
- Startup.cs

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #6926

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs [hunk 1]
```diff
@@ -42,5 +43,31 @@ public string RootDirectory
                 _root = value;
             }
         }
+
+        public bool DisableAreas { get; set; }
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: @Eilon \ @mkArtakMSFT naming suggestions

Comment (mkArtakMSFT) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: @pranavkm, RespectAreas or IgnoreAreas?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs [hunk 2]
```diff
@@ -17,14 +18,24 @@ public class CompiledPageRouteModelProvider : IPageRouteModelProvider
...
+        {
+        }
+
+        public CompiledPageRouteModelProvider(
+            ApplicationPartManager applicationManager,
+            IOptions<RazorPagesOptions> pagesOptionsAccessor,
+            ILoggerFactory loggerFactory)
         {
             _applicationManager = applicationManager;
             _pagesOptions = pagesOptionsAccessor.Value;
```
Comment (mkArtakMSFT) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: Not directly related to your changes, but "pagesOptionsAccessor" can be null, and that would throw here.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs [hunk 3]
```diff
@@ -30,20 +35,72 @@ public static void PopulateDefaults(PageRouteModel model, string routeTemplate)
                 // force outgoing routes to match to the path sans /Index.
                 selectorModel.AttributeRouteModel.SuppressLinkGeneration = true;
 
-                var parentDirectoryPath = model.ViewEnginePath;
-                var index = parentDirectoryPath.LastIndexOf('/');
+                string parentDirectoryPath;
+                var index = pageRoute.LastIndexOf('/');
                 if (index == -1)
```
Comment (mkArtakMSFT) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: Minor code reduction:
string parentDirectoryPath = index==-1?string.Empty:pageRoute.Substring(0, index);

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs (hunk 1 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: Spoke to @DamianEdwards and he suggested using `EnableAreas`.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs (hunk 2 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: We've historically not had null checks in ctors that are meant to be DI-ed and types not callable via user code.

**Comment by pranavkm** [comment]
🆙 📅

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesWithBasePathTest.cs [hunk 4]
```diff
@@ -280,13 +280,95 @@ public RazorPagesWithBasePathTest(MvcTestFixture<RazorPagesWebSite.StartupWithBa
...
+            // Arrange
+            var expected = 
+@"Layout in /Views/Shared
+Partial in /Areas/Accounts/Pages/Manage/
+
+Partial in /Areas/Accounts/Pages/
+
+Partial in /Areas/Accounts/Pages/
+
+Partial in /Areas/Accuonts/Views/Shared/
```
Comment (pranavkm) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/RazorPagesWithBasePathTest.cs]: typo

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs [hunk 5]
```diff
@@ -46,28 +47,52 @@ public class PageActionDescriptorChangeProvider : IActionDescriptorChangeProvide
...
             var importFileAtPagesRoot = rootDirectory + "/" + templateEngine.Options.ImportsFileName;
-            _additionalFilesToTrack = templateEngine.GetImportItems(importFileAtPagesRoot)
-                .Select(item => item.FilePath)
-                .ToArray();
+            var additionalImportFilePaths = templateEngine.GetImportItems(importFileAtPagesRoot)
+                .Select(item => item.FilePath);
+            var pagesRootSearchPattern = rootDirectory + "/**/*.cshtml";
 
-            _searchPattern = rootDirectory + "/**/*.cshtml";
+            if (razorPagesOptions.Value.EnableAreas)
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs]: You should do this by default and unconditionally. We don't have this feature for MVC and razor should be no different. What is your motivation/concern to add it?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs [hunk 6]
```diff
@@ -122,5 +128,13 @@ public static void MalformedPageDirective(this ILogger logger, string filePath,
...
             }
         }
+
+        public static void UnsupportedAreaPath(this ILogger logger, RazorPagesOptions options, string filePath)
+        {
+            if (logger.IsEnabled(LogLevel.Warning))
+            {
+                _unsupportedAreaPath(logger, filePath, options.AreaRootDirectory, options.AreaRootDirectory, options.RootDirectory, null);
+    }
+}
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs]: I think the diff got confused here, if not. Indentation?_

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs [hunk 7]
```diff
@@ -2,25 +2,30 @@
...
+using Microsoft.Extensions.Primitives;
 
 namespace Microsoft.AspNetCore.Mvc.RazorPages.Internal
 {
     public static class PageSelectorModel
     {
         private const string IndexFileName = "Index.cshtml";
 
-        public static void PopulateDefaults(PageRouteModel model, string routeTemplate)
+        public static void PopulateDefaults(PageRouteModel model, string pageRoute, string routeTemplate)
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: Breaking change?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs (hunk 5 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs]: Associating a controller with an area is opt-in via `AreaAttribute`, so it's not analogous to start with. In addition, we're adding a breaking behavior change to how the page gets routed to so there needs to be a way to disable the feature.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs (hunk 6 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs]: Probably a bad merge. Will fix.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs (hunk 7 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: It's an internal type. Don't think ApiCheck complained.

File: src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs [hunk 8]
```diff
@@ -471,6 +471,12 @@ public static string Page(this IUrlHelper urlHelper, string pageName, object val
                 routeValues["handler"] = pageHandler;
             }
 
+            if (!routeValues.ContainsKey("area") &&
+                ambientValues.TryGetValue("area", out var area))
+            {
+                routeValues["area"] = area;
+            }
+
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs]: This shouldn't be here, we don't special case `area` like we do for `[controller, action]` and `[page, handler]`

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.Core/UrlHelperExtensions.cs]: See https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Routing/UrlHelper.cs

File: src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngineOptions.cs [hunk 9]
```diff
@@ -112,12 +112,47 @@ public class RazorViewEngineOptions
...
+        ///
+        ///  /Areas/Accounts/Pages/Manage/{0}.cshtml
+        ///  /Areas/Accounts/Pages/{0}.cshtml
+        ///  /Areas/Accounts/Pages/Shared/{0}.cshtml
+        ///  /Areas/Accounts/Views/Shared/{0}.cshtml
+        ///  /Pages/Shared/{0}.cshtml
+        ///  /Views/Shared/{0}.cshtml
+        /// </para>
+        /// </remarks>
+        public IList<string> AreaPageViewLocationFormats { get; } = new List<string>();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/RazorViewEngineOptions.cs]: Solid 👍

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs [hunk 10]
```diff
@@ -52,6 +46,7 @@ public PageRouteModel(PageRouteModel other)
 
             Properties = new Dictionary<object, object>(other.Properties);
             Selectors = new List<SelectorModel>(other.Selectors.Select(m => new SelectorModel(m)));
+            RouteValues = new Dictionary<string, string>(other.RouteValues, StringComparer.OrdinalIgnoreCase);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs]: So in theory we probably should have had this here always right?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs [hunk 11]
```diff
@@ -73,5 +71,18 @@ public PageRouteModel(PageRouteModel other)
...
         public IList<SelectorModel> Selectors { get; }
+
+        /// <summary>
+        /// Gets a collection of route values that must be present in the <see cref="RouteData.Values"/> 
+        /// for the corresponding page to be selected.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// The value of <see cref="ViewEnginePath"/> is considered an implicit route value corresponding
+        /// to the key <c>page</c>. These entries will be added to <see cref="ActionDescriptor.RouteValues"/>,
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs]: `These entries will be added to` -> `These entries will be implicitly added to ... when the action descriptor is created...`

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs [hunk 12]
```diff
@@ -58,26 +69,63 @@ private void EnsureCache()
...
                 foreach (var viewDescriptor in GetViewDescriptors(_applicationManager))
                 {
-                    if (!viewDescriptor.RelativePath.StartsWith(rootDirectory, StringComparison.OrdinalIgnoreCase))
+                    PageRouteModel model = null;
+                    if (viewDescriptor.RelativePath.StartsWith(rootDirectory, StringComparison.OrdinalIgnoreCase))
                     {
-                        continue;
+                        model = GetPageRouteModel(rootDirectory, viewDescriptor);
+                    }
+                    else if (_pagesOptions.EnableAreas && viewDescriptor.RelativePath.StartsWith(areaRootDirectory, StringComparison.OrdinalIgnoreCase))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: Kinda interestingly this means that you can't have areas that differ only by case. I think it's probably fine, but worth bringing up. That makes sense to me because routing is already case-insensitive.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs [hunk 13]
```diff
@@ -58,26 +69,63 @@ private void EnsureCache()
...
+                {
+                    RouteValues = { ["area"] = result.areaName },
+                };
+                var pageAttribute = (RazorPageAttribute)viewDescriptor.ViewAttribute;
+                PageSelectorModel.PopulateDefaults(model, result.pageRoute, pageAttribute.RouteTemplate);
+
+                return model;
+            }
+
+            return null;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: What are some examples of this? I think it at least deserves a comment... Does it make sense for this to be an error? (I'm not sure what I think)

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs (hunk 5 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs]: Yeah I wish this were on by default, but it **is** a breaking change. We plan to introduce compatibility modes as well, which means that you can just set an app to `2.1 mode` and get all the new stuff.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs [hunk 14]
```diff
@@ -46,28 +47,52 @@ public class PageActionDescriptorChangeProvider : IActionDescriptorChangeProvide
...
+                    pagesRootSearchPattern,
+                    areaRootDirectory + "/**/*.cshtml",
+                };
+            }
+            else
+            {
+                _searchPatterns = new[] { pagesRootSearchPattern, };
+            }
+
+            _additionalFilesToTrack = additionalImportFilePaths.ToArray();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs]: Would it be possible to expose this and cover it in a unit test? I think it would be really helpful to see a test that shows exactly what gets tracked.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs [hunk 15]
```diff
@@ -55,6 +56,11 @@ static PageLoggerExtensions()
                 LogLevel.Trace,
                 2,
                 "{FilterType}: After executing {Method} on filter {Filter}.");
+
+            _unsupportedAreaPath = LoggerMessage.Define<string, string, string, string>(
+                LogLevel.Warning,
+                1,
+                "The page at '{FilePath}' is located under the area root directory '{AreaRootDirectory}' but does not follow the path format '{AreaRootDirectory}{RootDirectory}/Directory/FileName.cshtml");
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageLoggerExtensions.cs]: Ah, I see, yeah I think this is a good new thing to log 👍

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs [hunk 16]
```diff
@@ -30,20 +35,66 @@ public static void PopulateDefaults(PageRouteModel model, string routeTemplate)
...
+            if (string.Compare(path, areaEndIndex + 1, normalizedPagesRootDirectory, 0, normalizedPagesRootDirectory.Length, StringComparison.OrdinalIgnoreCase) != 0)
+            {
+                logger?.UnsupportedAreaPath(razorPagesOptions, path);
+                return false;
+            }
+
+            var areaName = path.Substring(1, areaEndIndex - 1);
+
+            var pagePathIndex = areaEndIndex + normalizedPagesRootDirectory.Length;
+            var pageExtensionIndex = path.LastIndexOf('.');
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: Should we look for `.cshtml` here? Would that make it simpler?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs [hunk 17]
```diff
@@ -30,20 +35,66 @@ public static void PopulateDefaults(PageRouteModel model, string routeTemplate)
...
+            var pageExtensionLength = pageExtensionIndex == -1 || pageExtensionIndex < pagePathIndex ?
+                0 :
+                path.Length - pageExtensionIndex;
+
+            // PageName
+            var pageName = path.Substring(pagePathIndex, path.Length - pagePathIndex - pageExtensionLength);
+
+            var builder = new InplaceStringBuilder(areaEndIndex + pageName.Length);
+            builder.Append(path, 0, areaEndIndex);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: If you're starting at zero isn't this just a string.Concat?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/RazorProjectPageRouteModelProvider.cs [hunk 18]
```diff
@@ -36,12 +36,21 @@ public void OnProvidersExecuted(PageRouteModelProviderContext context)
...
+                AddAreaPageModels(context);
+            }
+        }
+
+        private void AddPageModels(PageRouteModelProviderContext context)
         {
             foreach (var item in _project.EnumerateItems(_pagesOptions.RootDirectory))
             {
-                if (item.FileName.StartsWith("_"))
+                if (!IsRouteable(item))
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/RazorProjectPageRouteModelProvider.cs]: 👍

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/RazorProjectPageRouteModelProvider.cs [hunk 19]
```diff
@@ -51,35 +60,83 @@ public void OnProvidersExecuting(PageRouteModelProviderContext context)
...
-        private bool IsAlreadyRegistered(PageRouteModelProviderContext context, RazorProjectItem projectItem)
+        private bool IsAlreadyRegistered(PageRouteModelProviderContext context, PageRouteModel routeModel)
         {
             for (var i = 0; i < context.RouteModels.Count; i++)
             {
-                var routeModel = context.RouteModels[i];
-                if (string.Equals(routeModel.ViewEnginePath, projectItem.FilePathWithoutExtension, StringComparison.OrdinalIgnoreCase) &&
-                    string.Equals(routeModel.RelativePath, projectItem.CombinedPath, StringComparison.OrdinalIgnoreCase))
+                var existingRouteModel = context.RouteModels[i];
+                if (string.Equals(existingRouteModel.ViewEnginePath, routeModel.ViewEnginePath, StringComparison.OrdinalIgnoreCase) &&
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/RazorProjectPageRouteModelProvider.cs]: I don't think we really need to check both of these right? Either one of them is a source of truth

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageActionDescriptor.cs [hunk 20]
```diff
@@ -7,7 +7,7 @@
 
 namespace Microsoft.AspNetCore.Mvc.RazorPages
 {
-    [DebuggerDisplay("{" + nameof(ViewEnginePath) + "}")]
+    [DebuggerDisplay("{ViewEnginePath = " + nameof(ViewEnginePath) + ", RelativePath = " + nameof(RelativePath) + "}")]
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/PageActionDescriptor.cs]: You need to move the braces if you want to do this, they need to surround the `nameof()`s.

I'm personally super not a fan of nameof in these strings. it makes it an unreadable mess, and these names are public APIs and can't be changed anyway. You could also take the approach of moving the actually display to a method.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs [hunk 21]
```diff
@@ -42,5 +42,41 @@ public string RootDirectory
...
+        /// Gets or sets a value that determines if areas are enabled for Razor Pages.
+        /// Defaults to <c>true</c>.
+        /// <para>
+        /// When enabled, any Razor Page under the directory structure <c>/{AreaRootDirectory}/AreaName/{RootDirectory}/</c>
+        /// will be associated with an area with the name <c>AreaName</c>.
+        /// <seealso cref="AreaRootDirectory"/>
+        /// <seealso cref="RootDirectory"/>
+        /// </para>
+        /// </summary>
+        public bool EnableAreas { get; set; } = true;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: How did we decide on `true`? That's a breaking change

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs (hunk 14 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs]: I checked the tests, and this seems like it has enough coverage. I thought this code was a bit hard to follow, even as someone who already knows it pretty well. Consider cleaning this up or adding some explanations in comments

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs (hunk 21 above)
Comment (Eilon) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: I personally think it's a super minor breaking change. I could certainly be swayed about the default value depending on what we do regarding Compat Switches/Levels. But if this is off by default for at least *new* projects (created from templates), that would be a major usability problem. But I think I want it on even for "legacy" projects. I feel that the only scenario it breaks is something incredibly obscure.

Comment (Eilon) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: Actually I slightly take this back. I had confused myself over what scenario it breaks. It breaks the scenario where you set the Pages Root to be `/`, and I actually have no sense of whether that is common or not. I *hope* it's very uncommon, because it strikes me as a very dangerous thing to do. Razor Pages will find all kinds of crazy CSHTML files the developer left laying around their app (including MVC views!). So, it would easily break someone who did something I consider to be odd, but maybe that's ok?

Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: > I could certainly be swayed about the default value depending on what we do regarding Compat Switches/Levels. But if this is off by default for at least new projects 

The general principle of the compat switches design I've been working on is that new behaviours are on new projects created with the template.

The reason why I wanted to get compat switches into 2.1 is that it makes our decision process very mechanical from now on. I think we've got 5 of these things (changes in a shipped feature).

Comment (Eilon) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: Agreed.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs (hunk 10 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/ApplicationModels/PageRouteModel.cs]: Yup. We missed a couple of features in PageApplicatonModel \ PageRouteModel that controllers had.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs (hunk 12 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: It's already problematic with precompiled views since we don't allow two view paths to differ in case.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs (hunk 13 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: Sure. Most of it is stray Razor Page cshtml sitting in the Areas root e.g. `/Areas/Page.cshtml`  or `/Areas/MyArea/Page.cshtml`. We log warning for these as part of `TryParseAreaPath` but I'm not sure if that's enough.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs [hunk 22]
```diff
@@ -58,26 +69,63 @@ private void EnsureCache()
                     rootDirectory = rootDirectory + "/";
                 }
 
+                var areaRootDirectory = _pagesOptions.AreaRootDirectory;
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: Consider this scenario:

```
/Areas/First.cshtml
/Areas/MyArea/Pages/Second.cshtml
```

where `RootDirectory = /, AreaRootDirectory = /Areas`

Should we

1) Disallow overlapping of `RootDirectory` and `AreaRootDirectory`
2) Disallow `RootDirectory` based routes for things that match an area path. In this case, we'll end up with `/Areas/First` and `/MyArea/Second` as the routes
3) Don't do anything special for area paths. In this case, we'll end up with `/Areas/First`, `/Areas/MyArea/Pages/Second`, and `/MyArea/Second` as the routes

IMO, option 2, seems like the most appropriate behavior.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs (hunk 17 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: Doh.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/RazorProjectPageRouteModelProvider.cs (hunk 19 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/RazorProjectPageRouteModelProvider.cs]: Yup. I think I was being over cautious when I was writing this.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/PageActionDescriptor.cs (hunk 20 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/PageActionDescriptor.cs]: Sure

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs (hunk 21 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: Fair enough. If we do leave this not-enabled, @javiercn would have to set it as part of the Identity setup.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs (hunk 22 above)
Comment (javiercn) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: If I understand this correctly, you want to prevent people from havign a page like /Areas/First.cshtml? If that's the case, I'm ok with it. That said, I think I'm also fine not doing anything as you need to go out of your way to shoot yourself in the foot.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs (hunk 12 above)
Comment (javiercn) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: Yeah, I think it makes for a better experience to be case insensitive as having areas/files differ only in casing (and rely on that) is kinda edgy.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs (hunk 5 above)
Comment (javiercn) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs]: I chatted with @rynowak. While I agree this is a breaking change (as in a behavior change), it relies on two requirements that I don't think would be too common to happen at the same time:
* Change the pages root to be the content root.
* Have an existing Areas folder on your application with pages inside it.

There are a few options here:
* Do the "breaking change" and not provide any option.
* Do the "breaking change" by default and provide the ability to turn it off.
* Do not do the "breaking change" and provide the ability to turn the behavior on with the new compatibility switches functionality.

IMO I would just go ahead and do 1, any other solution adds complexity to this feature with very limited value (based on the assumption made above).

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs (hunk 7 above)
Comment (javiercn) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: Ah, yes. I believe this is OK in non-patch releases.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs (hunk 5 above)
Comment (DamianEdwards) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs]: I think we'll end up using the new compatibility modes to have our cake and eat it too. That is, this will be off in 2.0 compat mode, but on in 2.1+ compat mode, with ability to turn it on/off in either mode via the option.

**Comment by DamianEdwards** [comment]
This seems like more than 10 lines 😕

**Comment by pranavkm** [comment]
Sure it is in base 128.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs (hunk 21 above)
Comment (javiercn) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: Moving the comment above to here:

I chatted with @rynowak. While I agree this is a breaking change (as in a behavior change), it relies on two requirements that I don't think would be too common to happen at the same time:
* Change the pages root to be the content root.
* **Have an existing Areas folder on your application with pages inside it.**

There are a few options here:

* Do the "breaking change" and not provide any option.
* Do the "breaking change" by default and provide the ability to turn it off.
* Do not do the "breaking change" and provide the ability to turn the behavior on with the new compatibility switches functionality.

I would like to suggest that for preview 1 we do the feature on by default and without providing any option/compatibility switch for it, make an announcement on the announcements repository and gather feedback from people in order to make a decision that, absent data, adds complexity to this feature with limited value (based on the assumptions made above, we don't expect people to be broken by this)

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs (hunk 5 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageActionDescriptorChangeProvider.cs]: Fair enough. I'll keep the flag for now and file a work item to replace it once the compat switch feature lands.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs (hunk 17 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/PageSelectorModel.cs]: It used to be a bit messier which is why I thought I missed something obvious AFAIK, there isn't a way to concat a substring without calling `.Substring` that I wanted to avoid here.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/RazorProjectPageRouteModelProvider.cs (hunk 19 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/RazorProjectPageRouteModelProvider.cs]: Having the extra check does mean we don't have to do anything special to deal with the case of the two directories overlapping (since they have the same relative path but different view engine paths). Files in different areas could have the same view engine path, so that's not sufficient by itself.

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs [hunk 23]
```diff
@@ -17,14 +19,24 @@ public class CompiledPageRouteModelProvider : IPageRouteModelProvider
         private readonly object _cacheLock = new object();
         private readonly ApplicationPartManager _applicationManager;
         private readonly RazorPagesOptions _pagesOptions;
+        private readonly ILogger<CompiledPageRouteModelProvider> _logger;
         private List<PageRouteModel> _cachedModels;
 
         public CompiledPageRouteModelProvider(
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: Do we need this? Is this just for unit tests? Consider just nuking it. It's easy in tests to just use the null logger

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs (hunk 22 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: I spoke with @DamianEdwards and @Eilon  about this today. We agreed with 2.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs (hunk 21 above)
Comment (rynowak) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: Discussed @DamianEdwards and @Eilon - the plan is to make this off by default.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs (hunk 22 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/CompiledPageRouteModelProvider.cs]: Tracking via https://github.com/aspnet/Mvc/issues/7147

File: src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs (hunk 21 above)
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.RazorPages/RazorPagesOptions.cs]: Tracked via https://github.com/aspnet/Mvc/issues/7148
